rfd/0000-prototype-job-worker.md

```diff
@@ -0,0 +1,203 @@
+    - Processes are immediately killed, ensuring all resources are properly released.
+4. **Query Status**:
+    - The status of any job can be queried to return whether it is running,
+      stopped, or completed. For this initial prototype, the information provided
+      will be basic. However, a final product would expand on this to include
+      more detailed information, such as the return code, error output, and other
+      relevant diagnostics to provide a comprehensive view of the job's execution
+      status.
+
+#### Namespace Implementation
+Namespaces are a feature of the Linux kernel that provides isolation for
+running processes. They make it possible for processes to have their own view
+of the system's global resources, such as process IDs, network stacks, mount
+points, and interprocess communication resources. We will specifically use the
+following namespaces:
+  - **PID Namespace**: Ensures process isolation by providing
+    processes with separate PID namespaces. This prevents them from seeing or
+    interacting with processes in other namespaces.
+  - **Network Namespace**: Isolates network interfaces, IP routing tables, port
+    numbers, etc., allowing processes to operate on a network as if they were
+    the only ones on the system. To stay in scope, the network will remain
+    isolated.
+  - **Mount Namespace**: Isolates filesystem mount points, allowing processes
+    to have their own set of filesystems visible to them.
+
+##### Utilization Strategy
+Our service will utilize the syscall package in Go to unshare and manage
+namespaces. The following code snippet demonstrates how a new process can be
+started in a new set of namespaces:
+
+```go
+import "syscall"
+
+func setupNamespaces() error {
+    // Unshare creates new namespaces for the calling process
+    if err := syscall.Unshare(syscall.CLONE_NEWPID | syscall.CLONE_NEWNET | syscall.CLONE_NEWNS); err != nil {
+        return fmt.Errorf("failed to create namespaces: %v", err)
+    }
+    return nil
+}
+```
+
+#### Cgroup Implementation
+Control groups (cgroups) are another kernel feature that limits, accounts for,
+and isolates the resource usage (CPU, memory, disk I/O, etc.) of process
+groups. We will implement cgroups to ensure that each job runs within
+predefined resource limits, improving system stability and preventing any job
+from monopolizing system resources.
+
+##### Configuration Approach
+Cgroups will be managed by interacting with the cgroup filesystem typically
+mounted under /sys/fs/cgroup. Here is how we plan to create and manage cgroups:
+
+```go
+import (
+    "io/ioutil"
+    "path/filepath"
+)
+
+func createCgroup(path, resource, limit string) error {
+    cgroupPath := filepath.Join("/sys/fs/cgroup", resource, path)
+    if err := os.MkdirAll(cgroupPath, 0755); err != nil {
+        return err
+    }
+    return ioutil.WriteFile(filepath.Join(cgroupPath, "tasks"), []byte(strconv.Itoa(os.Getpid())), 0644)
```

Comment:
Going with `v2`, it will limit the backwards compatibility, but `v2` has been around for long enough that I feel comfortable just documenting this limitation.